File: 2015/14.jl
mutable struct Reindeer
    name::String
    speed::Int32
    energy::Int32
    recover::Int32
    km::Int32
    state::Int16
    points::Int32
end

function Reindeer(n, s, e, r)
    return Reindeer(n, s, e, r, 0, 0, 0)
end

function print(x::Reindeer)
    println(x.name, " - ", x.km, " - ", x.points)
end

function run(x::Reindeer)
    if x.state < 0
        x.state += 1
        if x.state == 0
            x.state = x.energy
        end
    else
        x.state -= 1
        x.km += x.speed
        if x.state == 0
            x.state = -x.recover
        end
    end
end

function scoring(reindeers::Vector{Reindeer})
    m = maximum([r.km for r in reindeers])
    indices = findall(x->x.km == m, reindeers)
    for i in indices
        reindeers[i].points += 1
    end

    return reindeers
end

function readInput()
    reindeers = Vector{Reindeer}()
    f = open("instructions14a.txt", "r")
    lines = readlines(f)
    for l in lines
        d = split(l, " ")
        r = Reindeer(d[1], parse(Int, d[4]), parse(Int, d[7]), parse(Int, d[end-1]))
        push!(reindeers, r)
        #println(reindeers[end])
    end       
    return reindeers
end

function part1(reindeers, time)    
    for t in 1:time+1
        for r in reindeers
            run(r)
        end 
    end 

    kms = maximum([r.km for r in reindeers])
    winner = findfirst(x->x.km==kms, reindeers)
    println("🎅 ", reindeers[winner])
end

function part2(reindeers, time)    
    for t in 1:time+1
        for r in reindeers
            run(r)
        end 
        reindeers = scoring(reindeers)
    end 
    
    for r in reindeers
        print(r)
    end
end

time = 100
r = readInput()
part1(r, time)
r = readInput()
part2(r, time)

File: 2015/14.py
import time

from utils import readInput

class Reindeer:
    def __init__(self, name, speed, energy, recover):
        self.name = name
        self.speed = speed
        self.energy = energy
        self.recover = recover
        self.km = 0
        self.state = self.energy
        self.points = 0

    def __repr__(self):
        return (self.name + " " + str(self.km) + " " + str(self.points))

    def run(self):
        if self.state < 0:
            self.state += 1
            if self.state == 0:
                self.state = self.energy
        else:
            self.state -= 1
            self.km += self.speed
            if self.state == 0:
                self.state = -self.recover

def loadInput():
    lines = readInput("instructions14a.txt")
    reindeers = []
    for d in lines:
        d = d.split("\n")[0].split(" ")
        reindeers.append(Reindeer(d[0], int(d[3]), int(d[6]), int(d[-2])))
    return reindeers

def scoring(reindeers):
    kms = []
    for r in reindeers:
        kms.append(r.km)

    m = max(kms)
    for i, k in enumerate(kms):
        if k == m:
            reindeers[i].points += 1
    return reindeers

def run(reindeers):
    time = 2503

    for t in range(time):
        for r in reindeers:
            r.run()
        reindeers = scoring(reindeers)

def part1(reindeers):
    run(reindeers)
    reindeers.sort(key=lambda r: r.km, reverse=True)
    print (f"🎄 Part 1: {reindeers[0]}")
    return reindeers

def part2(reindeers):
    reindeers.sort(key=lambda r: r.points, reverse=True)
    print (f"🎄🎅 Part 2: {reindeers[0]}")
    
if __name__ == "__main__":
    title = "Day 14: Reindeer Olympics"
    sub = "-"*(len(title)+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput()
    
    t0 = time.time()
    reindeers = part1(inputs)
    print ("Time: {:.5f}".format(time.time()-t0))
    
    t0 = time.time()
    part2(reindeers)
    print ("Time: {:.5f}".format(time.time()-t0))

